File: qviability/export/pdf.py
"""QC PDF export — one page per image, the original beside the marked version.

A flip-through visual record for final quality control: on each page the
reviewer sees the plain image ("Before") and exactly which pixels the chosen
threshold counted ("After"), side by side. The rendering reuses the same
display pipeline as the annotated PNGs, so the PDF matches what was exported.
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
from PIL import Image, ImageDraw

from ..core import image_io
from ..core.project import ImageEntry, Project
from . import annotate

# Page geometry, in pixels (saved at 150 DPI -> ~10 x 7 inch landscape).
PAGE_W, PAGE_H = 1500, 1060
MARGIN = 40
GUTTER = 28
HEADER_H = 84               # room for the two header text lines
LABEL_H = 32                # room for the "Before" / "After" panel labels

PAGE_BG = (255, 255, 255)
PANEL_BG = (24, 24, 24)
PANEL_BORDER = (170, 170, 170)
INK = (20, 20, 20)
SUBINK = (95, 95, 95)


def _fmt_int(n: float) -> str:
    return f"{round(n):,}"


def _panel(rgb: np.ndarray, box_w: int, box_h: int) -> Image.Image:
    """Scale *rgb* to fit a box (aspect preserved) on a dark panel background."""
    panel = Image.new("RGB", (box_w, box_h), PANEL_BG)
    img = Image.fromarray(np.ascontiguousarray(rgb))
    img.thumbnail((box_w, box_h), Image.Resampling.LANCZOS)
    panel.paste(img, ((box_w - img.width) // 2, (box_h - img.height) // 2))
    return panel


def render_qc_page(project: Project, entry: ImageEntry) -> Image.Image | None:
    """Render one QC page for *entry*.

    Returns None if the image cannot be read or *entry* has no threshold yet.
    """
    if entry.threshold is None:
        return None
    try:
        raw = image_io.get_processed(entry.path, project.subtract_background,
                                     project.radius_for(entry.channel))
    except Exception:                                  # noqa: BLE001
        return None

    before = annotate.render_display_rgb(
        raw, entry.brightness, entry.contrast)
    after = annotate.render_overlay_rgb(
        raw, entry.threshold, entry.channel,
        entry.brightness, entry.contrast, entry.exclusions)

    page = Image.new("RGB", (PAGE_W, PAGE_H), PAGE_BG)
    draw = ImageDraw.Draw(page)

    # Header --------------------------------------------------------------
    title_font = annotate._load_font(30)
    sub_font = annotate._load_font(20)
    label_font = annotate._load_font(22)

    draw.text((MARGIN, MARGIN), entry.display_name, fill=INK, font=title_font)

    m = entry.measurement
    parts = [entry.channel, f"threshold {entry.threshold}"]
    if m:
        parts.append(f"positive area {m.positive_area_pct:.2f}%")
        parts.append(f"integrated intensity {_fmt_int(m.integrated_intensity)}")
    if project.subtract_background:
        parts.append("background subtracted")
    draw.text((MARGIN, MARGIN + 40), "     ".join(parts),
              fill=SUBINK, font=sub_font)

    # Two panels ----------------------------------------------------------
    panel_top = MARGIN + HEADER_H + LABEL_H
    panel_h = PAGE_H - panel_top - MARGIN
    panel_w = (PAGE_W - 2 * MARGIN - GUTTER) // 2

    for i, (label, rgb) in enumerate((("Before  (original)", before),
                                      ("After  (marked = counted)", after))):
        x0 = MARGIN + i * (panel_w + GUTTER)
        draw.text((x0, MARGIN + HEADER_H), label, fill=INK, font=label_font)
        page.paste(_panel(rgb, panel_w, panel_h), (x0, panel_top))
        draw.rectangle([x0, panel_top, x0 + panel_w - 1,
                        panel_top + panel_h - 1], outline=PANEL_BORDER, width=1)

    return page


def save_pdf(pages: list[Image.Image], out_path: str) -> None:
    """Write *pages* as a single multi-page PDF at *out_path*."""
    out = Path(out_path)
    out.parent.mkdir(parents=True, exist_ok=True)
    pages[0].save(str(out), "PDF", save_all=True,
                  append_images=pages[1:], resolution=150.0)

File: web/src/export/pdf.ts
// QC PDF export — port of qviability/export/pdf.py.
//
// One landscape page per tuned image: the plain image ("Before") beside the
// marked version ("After"), with a header and panel labels. Built with pdf-lib
// using a bundled DejaVuSans font (so glyphs render without relying on viewer
// fonts). Before/after panels reuse the float-blend renderer in annotate.ts.

import { PDFDocument, rgb, type PDFFont, type PDFImage } from "pdf-lib";
import fontkit from "@pdf-lib/fontkit";
import type { ImageEntry, Project, RawImage } from "@/core/types";
import {
  renderDisplayRgb,
  renderOverlayRgb,
  rgbToCanvas,
  canvasToPngBlob,
  type RgbImage,
} from "./annotate";

const PAGE_W = 1500;
const PAGE_H = 1060;
const MARGIN = 40;
const GUTTER = 28;
const HEADER_H = 84;
const LABEL_H = 32;

const INK = rgb(20 / 255, 20 / 255, 20 / 255);
const SUBINK = rgb(95 / 255, 95 / 255, 95 / 255);
const PANEL_BG = rgb(24 / 255, 24 / 255, 24 / 255);
const PANEL_BORDER = rgb(170 / 255, 170 / 255, 170 / 255);

export interface QcItem {
  entry: ImageEntry;
  raw: RawImage;
}

function fmtInt(n: number): string {
  return Math.round(n).toLocaleString("en-US");
}

/** Downscale (never upscale) an RgbImage to fit box, return PNG bytes + draw size. */
async function panelPng(
  img: RgbImage,
  boxW: number,
  boxH: number,
): Promise<{ bytes: Uint8Array; w: number; h: number }> {
  const scale = Math.min(1, boxW / img.width, boxH / img.height);
  const w = Math.max(1, Math.round(img.width * scale));
  const h = Math.max(1, Math.round(img.height * scale));
  const src = rgbToCanvas(img);
  const dst = document.createElement("canvas");
  dst.width = w;
  dst.height = h;
  const ctx = dst.getContext("2d")!;
  ctx.imageSmoothingEnabled = true;
  ctx.imageSmoothingQuality = "high";
  ctx.drawImage(src, 0, 0, w, h);
  const blob = await canvasToPngBlob(dst);
  return { bytes: new Uint8Array(await blob.arrayBuffer()), w, h };
}

async function loadFontBytes(): Promise<ArrayBuffer> {
  const url = `${import.meta.env.BASE_URL}fonts/DejaVuSans.ttf`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Could not load PDF font (${res.status})`);
  return res.arrayBuffer();
}

async function renderPage(
  doc: PDFDocument,
  font: PDFFont,
  project: Project,
  item: QcItem,
): Promise<void> {
  const { entry, raw } = item;
  const before = renderDisplayRgb(raw, entry.brightness, entry.contrast);
  const after = renderOverlayRgb(
    raw,
    entry.threshold ?? 0,
    entry.channel,
    entry.brightness,
    entry.contrast,
    entry.exclusions,
  );

  const page = doc.addPage([PAGE_W, PAGE_H]);
  // White background.
  page.drawRectangle({ x: 0, y: 0, width: PAGE_W, height: PAGE_H, color: rgb(1, 1, 1) });

  const top = (yFromTop: number) => PAGE_H - yFromTop;

  // Header.
  page.drawText(entry.display_name, {
    x: MARGIN,
    y: top(MARGIN + 30),
    size: 30,
    font,
    color: INK,
  });
  const m = entry.measurement;
  const parts = [entry.channel, `threshold ${entry.threshold}`];
  if (m) {
    parts.push(`positive area ${m.positive_area_pct.toFixed(2)}%`);
    parts.push(`integrated intensity ${fmtInt(m.integrated_intensity)}`);
  }
  if (project.subtract_background) parts.push("background subtracted");
  page.drawText(parts.join("     "), {
    x: MARGIN,
    y: top(MARGIN + 40 + 20),
    size: 20,
    font,
    color: SUBINK,
  });

  const panelTop = MARGIN + HEADER_H + LABEL_H;
  const panelH = PAGE_H - panelTop - MARGIN;
  const panelW = Math.floor((PAGE_W - 2 * MARGIN - GUTTER) / 2);

  const labels = ["Before  (original)", "After  (marked = counted)"];
  const imgs = [before, after];
  for (let i = 0; i < 2; i++) {
    const x0 = MARGIN + i * (panelW + GUTTER);
    page.drawText(labels[i], {
      x: x0,
      y: top(MARGIN + HEADER_H + 22),
      size: 22,
      font,
      color: INK,
    });
    // Panel background.
    page.drawRectangle({
      x: x0,
      y: top(panelTop + panelH),
      width: panelW,
      height: panelH,
      color: PANEL_BG,
    });
    // Scaled image, centered.
    const { bytes, w, h } = await panelPng(imgs[i], panelW, panelH);
    const png: PDFImage = await doc.embedPng(bytes);
    const ox = (panelW - w) / 2;
    const oy = (panelH - h) / 2;
    page.drawImage(png, {
      x: x0 + ox,
      y: top(panelTop + oy + h),
      width: w,
      height: h,
    });
    // Border.
    page.drawRectangle({
      x: x0,
      y: top(panelTop + panelH),
      width: panelW,
      height: panelH,
      borderColor: PANEL_BORDER,
      borderWidth: 1,
    });
  }
}

/** Build the multi-page QC PDF. Returns null when there are no tuned items. */
export async function buildPdf(
  project: Project,
  items: QcItem[],
): Promise<Uint8Array | null> {
  if (items.length === 0) return null;
  const doc = await PDFDocument.create();
  doc.registerFontkit(fontkit);
  const font = await doc.embedFont(await loadFontBytes());
  for (const item of items) {
    if (item.entry.threshold === null) continue;
    await renderPage(doc, font, project, item);
  }
  return doc.save();
}
